collect_app/src/main/java/org/odk/collect/android/widgets/GeoTraceWidget.java

```diff
@@ -144,7 +147,16 @@ public void setOnLongClickListener(OnLongClickListener l) {
 
     @Override
     public void onButtonClick(int buttonId) {
-        waitForData();
-        startGeoTraceActivity();
+        requestLocationPermissions((FormEntryActivity) getContext(), new PermissionListener() {
```

Comment:
Would it be better to check first whether permission has already been granted with 
`checkIfLocationPermissionsGranted` ?

Reply:
We use a library to simplify the whole work https://github.com/Karumi/Dexter
It works out of the box.